### PR TITLE
Add dockershim faq

### PR DESCRIPTION
**Pull request by mrbobbytables**
Adds faq regarding dockershim deprecation

### CODE REVIEW

File: content/en/docs/setup/production-environment/dockershim.md [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+For instructions on how to use containerd and CRI-O with Kubernetes, see the
+Kubernetes documentation on [Container Runtimes]
+
+[Container Runtimes]: /docs/setup/production-environment/container-runtimes
+
+### Why is dockershim being deprecated?
+Maintaining dockershim has become a heavy burden on the Kubernetes maintainers.
+The CRI standard was created to reduce this burden and allow smooth interoperability
+of different container runtimes, however Docker itself does not currently implement
+CRI. Dockershim was always intended to be a temporary solution (hence the name).
```
Comment (jeefy) [inline content/en/docs/setup/production-environment/dockershim.md:128]: "hence the name: shim"

Just being explicit

File: content/en/docs/setup/production-environment/dockershim.md [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+For instructions on how to use containerd and CRI-O with Kubernetes, see the
+Kubernetes documentation on [Container Runtimes]
+
+[Container Runtimes]: /docs/setup/production-environment/container-runtimes
+
+### Why is dockershim being deprecated?
+Maintaining dockershim has become a heavy burden on the Kubernetes maintainers.
+The CRI standard was created to reduce this burden and allow smooth interoperability
+of different container runtimes, however Docker itself does not currently implement
```
Comment (jeefy) [inline content/en/docs/setup/production-environment/dockershim.md:127]: runtimes. Docker itself doesn't currently implement CRI, thus the problem.

File: content/en/docs/setup/production-environment/dockershim.md [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+[Dockershim Removal Kubernetes Enhancement Proposal][drkep]
+
+[drkep]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1985-remove-dockershim
+
+
+### What if I have more questions?
+
+If you use a vendor-supported Kubernetes distribution, you can ask them about
+upgrade plans for their products. For end-user questions, you can find our
+support volunteers on https://discuss.kubernetes.io/.
```
Comment (jeefy) [inline content/en/docs/setup/production-environment/dockershim.md:139]: For end-user questions, you can ask the community questions on https://discuss.kubernetes.io/.

**Review by jeefy — CHANGES_REQUESTED** [review]
Minor changes that didn't carry over from the doc, the rest got in :D Thanks!

**Comment by mrunalp** [comment]
Reviewing...

File: content/en/docs/setup/production-environment/dockershim.md [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+transition and will evaluate things as the situation evolves. 
+
+### Are there examples of folks using other runtimes in production today that I can use to make the case to move forward with this change?
+
+The [kind] project has been using containerd for some time and has seen an
+improvement in stability for its use case. Kind and containerd are leveraged
+multiple times every day to validate any changes to the Kubernetes codebase. Other
+related projects follow a similar pattern as well, demonstrating the stability and
+usability of other container runtimes.
+
```
Comment (mrunalp) [inline content/en/docs/setup/production-environment/dockershim.md:34]: We can add a sentence around OpenShift 4.x using CRI-O in production since June 2019 as well here.

File: content/en/docs/setup/production-environment/dockershim.md [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+Docker. Standards like OCI and CRI have helped many tools grow and thrive in our
+ecosystem, some replacing aspects of Docker while others enhance existing
+functionality.
+
+
+### Which CRI implementation should I use?
+
+That’s a complex question and it depends on a lot of factors. If Docker is
+working for you, moving to containerd should be a relatively easy swap and
+has have strictly better performance and le overhead. However we encourage you
```
Comment (mrunalp) [inline content/en/docs/setup/production-environment/dockershim.md:79]: le -> less

File: content/en/docs/setup/production-environment/dockershim.md [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+Maintaining dockershim has become a heavy burden on the Kubernetes maintainers.
+The CRI standard was created to reduce this burden and allow smooth interoperability
+of different container runtimes, however runtimes. Docker itself doesn't currently
+implement CRI, thus the problem. Dockershim was always intended to be a temporary
+solution (hence the name: shim). You can read more about the community discussion
+and planning in the [Dockershim Removal Kubernetes Enhancement Proposal][drkep]
+
+[drkep]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1985-remove-dockershim
+
```
Comment (mrunalp) [inline content/en/docs/setup/production-environment/dockershim.md:138]: Newer features such as cgroups v2 and user namespaces are being implemented in the new CRI runtimes and cutting away from dockershim allows us to move faster in supporting new kernel features for containerization.

File: content/en/docs/setup/production-environment/dockershim.md [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+[cr]: https://github.com/kubernetes-sigs/cri-tools
+[img]: https://github.com/genuinetools/img
+[buildah]: https://github.com/containers/buildah
+[kaniko]: https://github.com/GoogleContainerTools/kaniko
+
+For containerd in particular, you can start with their [documentation] to see what
+configuration options are available as you migrate things over.
+
+[documentaiton]: https://github.com/containerd/cri/blob/master/docs/registry.md
```
Comment (mrunalp) [inline content/en/docs/setup/production-environment/dockershim.md:120]: typo: documentation

File: content/en/docs/setup/production-environment/dockershim.md [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+---
+
+This document goes over some frequently asked questions and their answers. For
+more detail on the deprecation of Docker as a container runtime for Kubernetes
+kubelets, and what that means, check out the blog post
+[Don't Panic: Kubernetes and Docker](/blog/2020/12/02/dont-panic-kubernetes-and-docker/)
+
+### Can I still use Docker in Kubernetes 1.20?
+
+Yes, the only thing changing in 1.20 is a single warning log printed at startup
```
Comment (BenTheElder) [inline content/en/docs/setup/production-environment/dockershim.md:15]: should we mention that it's in kubelet startup, specifically? (and maybe link to the docs for kubelet?)

https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/

File: content/en/docs/setup/production-environment/dockershim.md [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+CRIs (including containerd), there are a few differences around the edges. Some
+common things to consider when migrating are:
+- Logging configuration
+- Runtime resource limitations
+- Node provisioning scripts that require Docker or the controlsocket
+- Kubectl plugins that require docker CLI or the control socket
+- Kubernetes tools that require direct access to Docker (e.g. kube-imagepuller)
+- Configuration of functionality like `registry-mirrors` and insecure registries 
+- Other support scripts or daemons that expect docker to be available and are run
+  outside of Kubernetes (e.g. monitoring or security agents)
```
Comment (mrunalp) [inline content/en/docs/setup/production-environment/dockershim.md:98]: We can another bullet for how devices are enabled in runtimes so users that want GPUs or other special hardware should look up the documentation on how to enable those when switching runtimes.

File: content/en/docs/setup/production-environment/dockershim.md [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+Also yes. All CRI runtimes support the same pull secrets configuration used in
+Kubernetes, either via the PodSpec or ServiceAccount.
+
+
+### People keep referencing OCI, what is that?
+
+OCI stands for the [Open Container Initiative], which standardized a lot of the
+interfaces between container tools and technologies. They maintain a standard
+specification for packaging container images (OCI image-spec) and running containers
+(OCI runtime-spec). The CRI builds on these low-level specifications to provide
```
Comment (mrunalp) [inline content/en/docs/setup/production-environment/dockershim.md:59]: We should also mention runc which is being maintained as an implementation of the runtime spec within OCI and is typically the low level runtimes that is the default in containerd and CRI-O as it was in docker as well.

File: content/en/docs/setup/production-environment/dockershim.md [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+
+
+### What should I look out for when changing CRI implementations?
+
+While the underlying containerization code is the same between Docker and most
+CRIs (including containerd), there are a few differences around the edges. Some
+common things to consider when migrating are:
+- Logging configuration
+- Runtime resource limitations
+- Node provisioning scripts that require Docker or the controlsocket
```
Comment (mrunalp) [inline content/en/docs/setup/production-environment/dockershim.md:93]: .. call docker or use docker via it's control socket

**Comment by mrbobbytables** [comment]
okay, I think all changes have now been made

**Comment by jeefy** [comment]
lgtm :taco: Won't explicitly lgtm this one though.

File: content/en/docs/setup/production-environment/dockershim.md [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+Yes, the only thing changing in 1.20 is a single warning log printed at [kubelet]
+startup if using Docker as the runtime.
+
+[kubelet]: /docs/reference/command-line-tools-reference/kubelet/
+
+### When will dockershim be removed?
+
+Given the impact of this change, we are using an extended deprecation timeline.
+It will not be removed before Kubernetes 1.22, meaning the earliest release without
+dockershim would be 1.23 in late 2021, approximately one year from now. We will
```
Comment (coderanger) [inline content/en/docs/setup/production-environment/dockershim.md:24]: I would remove the "approximately one year from now" because this won't have a date stamp on it.

Comment (justaugustus) [reply content/en/docs/setup/production-environment/dockershim.md:24]: Agreed, especially as we're discussing changes to release cadence.

File: content/en/docs/setup/production-environment/dockershim.md [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+be working closely with vendors and other ecosystem groups to ensure a smooth
+transition and will evaluate things as the situation evolves. 
+
+### Are there examples of folks using other runtimes in production today that I can use to make the case to move forward with this change?
+
+The [kind] project has been using containerd for some time and has seen an
+improvement in stability for its use case. Kind and containerd are leveraged
+multiple times every day to validate any changes to the Kubernetes codebase. Other
```
Comment (justaugustus) [inline content/en/docs/setup/production-environment/dockershim.md:32]: Maybe we want to explicitly say that it's part of release signal?

File: content/en/docs/setup/production-environment/dockershim.md [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+title: Dockershim Deprecation FAQ
+content_type: concept
+aliases: [ '/dockershim']
+weight: 40
+---
+
+This document goes over some frequently asked questions and their answers. For
+more detail on the deprecation of Docker as a container runtime for Kubernetes
+kubelets, and what that means, check out the blog post
+[Don't Panic: Kubernetes and Docker](/blog/2020/12/02/dont-panic-kubernetes-and-docker/)
```
Comment (celestehorgan) [inline content/en/docs/setup/production-environment/dockershim.md:11]:
```suggestion
[Don't Panic: Kubernetes and Docker](/blog/2020/12/02/dont-panic-kubernetes-and-docker/).
```

**Review by justaugustus — COMMENTED** [review]
+1 to @coderanger's review comment and one other suggestion.

Outside of that, you've got an /approve from SIG Release leadership! Thanks for getting this out!

cc: @kubernetes/sig-release

File: content/en/docs/setup/production-environment/dockershim.md (hunk 13 above)
Comment (mrbobbytables) [reply content/en/docs/setup/production-environment/dockershim.md:32]: I don't think most end users will really get "release signal"

**Comment by mrunalp** [comment]
Thanks! /lgtm

File: content/en/docs/setup/production-environment/dockershim.md (hunk 13 above)
Comment (justaugustus) [reply content/en/docs/setup/production-environment/dockershim.md:32]: Something along the lines of "we check this before we released Kubernetes".

i.e., it's Community-validated for the artifacts we produce is a stronger message over it only being "leveraged".

File: content/en/docs/setup/production-environment/dockershim.md [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+For other examples and references you can look at the adopters of containerd and
+cri-o, two container runtimes under the Cloud Native Computing Foundation ([CNCF]).
+- [containerd](https://github.com/containerd/containerd/blob/master/ADOPTERS.md)
+- [CRI-O](https://github.com/cri-o/cri-o/blob/master/ADOPTERS.md)
+
+[CRI-O]: https://cri-o.io/
+[kind]: https://kind.sigs.k8s.io/
+[CNCF]: https://cncf.io
+
+### Will my existing Docker images still work?
```
Comment (celestehorgan) [inline content/en/docs/setup/production-environment/dockershim.md:46]:
```suggestion
### Will my existing Docker images still work?

```
header formatting breaks otherwise

File: content/en/docs/setup/production-environment/dockershim.md [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+
+### What about private images?
+
+Also yes. All CRI runtimes support the same pull secrets configuration used in
+Kubernetes, either via the PodSpec or ServiceAccount.
+
+
+### People keep referencing OCI, what is that?
+
+OCI stands for the [Open Container Initiative], which standardized a lot of the
```
Comment (celestehorgan) [inline content/en/docs/setup/production-environment/dockershim.md:59]:
```suggestion
OCI stands for the [Open Container Initiative], which standardized many of the
```

File: content/en/docs/setup/production-environment/dockershim.md [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+
+[Open Container Initative]: https://opencontainers.org/about/overview/
+[runc]: https://github.com/opencontainers/runc
+[containerd]: https://containerd.io/
+
+
+### Are Docker and containers the same thing?
+
+Docker popularized the Linux containers pattern and has been instrumental in
+developing the underlying technology, however containers in linux have existed
```
Comment (celestehorgan) [inline content/en/docs/setup/production-environment/dockershim.md:75]:
```suggestion
developing the underlying technology, however containers in Linux have existed
```

File: content/en/docs/setup/production-environment/dockershim.md [hunk 18]
```diff
@@ -0,0 +1,170 @@
+---
+title: Dockershim Deprecation FAQ
+content_type: concept
+aliases: [ '/dockershim']
+weight: 40
+---
+
+This document goes over some frequently asked questions and their answers. For
+more detail on the deprecation of Docker as a container runtime for Kubernetes
```
[the remark this answers is not in the document]
Comment (BenTheElder) [reply content/en/docs/setup/production-environment/dockershim.md:9]: there's already a blog post verison https://kubernetes.io/blog/2020/12/02/dont-panic-kubernetes-and-docker/

File: content/en/docs/setup/production-environment/dockershim.md [hunk 19]
```diff
@@ -0,0 +1,170 @@
+---
+title: Dockershim Deprecation FAQ
+content_type: concept
+aliases: [ '/dockershim']
+weight: 40
+---
+
+This document goes over some frequently asked questions and their answers. For
```
Comment (kbhawkey) [inline content/en/docs/setup/production-environment/dockershim.md:8]: This document reviews some frequently asked questions about [what]?

File: content/en/docs/setup/production-environment/dockershim.md [hunk 20]
```diff
@@ -0,0 +1,170 @@
...
+even better fit for your environment.
+
+[CNCF landscape]: https://landscape.cncf.io/category=container-runtime&format=card-mode&grouping=category
+
+
+### What should I look out for when changing CRI implementations?
+
+While the underlying containerization code is the same between Docker and most
+CRIs (including containerd), there are a few differences around the edges. Some
+common things to consider when migrating are:
```
Comment (celestehorgan) [inline content/en/docs/setup/production-environment/dockershim.md:97]:
```suggestion
common things to consider when migrating are:

```

(again, minor md formatting issue)

File: content/en/docs/setup/production-environment/dockershim.md [hunk 21]
```diff
@@ -0,0 +1,170 @@
...
+
+This document goes over some frequently asked questions and their answers. For
+more detail on the deprecation of Docker as a container runtime for Kubernetes
+kubelets, and what that means, check out the blog post
+[Don't Panic: Kubernetes and Docker](/blog/2020/12/02/dont-panic-kubernetes-and-docker/)
+
+### Can I still use Docker in Kubernetes 1.20?
+
+Yes, the only thing changing in 1.20 is a single warning log printed at [kubelet]
+startup if using Docker as the runtime.
```
Comment (kbhawkey) [inline content/en/docs/setup/production-environment/dockershim.md:16]: nit: Could you write the links as `[kubelet](/docs/reference/command-line-tools-reference/kubelet/)`

File: content/en/docs/setup/production-environment/dockershim.md [hunk 22]
```diff
@@ -0,0 +1,170 @@
...
+
+[documentation]: https://github.com/containerd/cri/blob/master/docs/registry.md
+
+For instructions on how to use containerd and CRI-O with Kubernetes, see the
+Kubernetes documentation on [Container Runtimes]
+
+[Container Runtimes]: /docs/setup/production-environment/container-runtimes
+
+
+### Why is dockershim being deprecated?
```
Comment (celestehorgan) [inline content/en/docs/setup/production-environment/dockershim.md:136]: Move this question to the very top of the list.

In short: order this based on **relevance to the casual passer-by**.  

If I know nothing about this change other than some scary twitter thread by a Senior Clout Engineer, I'd want the 5 W's (Who/What/When/Where/**Why**) answered before I dove into implementation details, picking a replacement, etc.

File: content/en/docs/setup/production-environment/dockershim.md [hunk 23]
```diff
@@ -0,0 +1,170 @@
...
+
+Maintaining dockershim has become a heavy burden on the Kubernetes maintainers.
+The CRI standard was created to reduce this burden and allow smooth interoperability
+of different container runtimes, however runtimes. Docker itself doesn't currently
```
Comment (celestehorgan) [inline content/en/docs/setup/production-environment/dockershim.md:140]:
```suggestion
of different container runtimes. Docker itself doesn't currently
```

side note: "However, runtimes" is my new objection to any Kubernetes enhancement. I don't comment on many keps tho so maybe @justaugustus can do the honors 😂

File: content/en/docs/setup/production-environment/dockershim.md [hunk 24]
```diff
@@ -0,0 +1,170 @@
...
+implement CRI, thus the problem. Dockershim was always intended to be a temporary
```
Comment (celestehorgan) [inline content/en/docs/setup/production-environment/dockershim.md:141]:
```suggestion
implement CRI, thus the problem. 

Dockershim was always intended to be a temporary
```

New paragraph here.

File: content/en/docs/setup/production-environment/dockershim.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (mrbobbytables) [reply content/en/docs/setup/production-environment/dockershim.md:9]: its being migrated to a blog post now. 👍

[the remark this answers is not in the document]
Comment (kbhawkey) [reply content/en/docs/setup/production-environment/dockershim.md:9]: Yes, I understand. How will readers know to look in this section for important information?
It would be good to broadcast widely and often (release notes?).

File: content/en/docs/setup/production-environment/dockershim.md [hunk 25]
```diff
@@ -0,0 +1,170 @@
...
+
+### When will dockershim be removed?
+
+Given the impact of this change, we are using an extended deprecation timeline.
+It will not be removed before Kubernetes 1.22, meaning the earliest release without
+dockershim would be 1.23 in late 2021, approximately one year from now. We will
+be working closely with vendors and other ecosystem groups to ensure a smooth
+transition and will evaluate things as the situation evolves. 
+
+### Are there examples of folks using other runtimes in production today that I can use to make the case to move forward with this change?
```
Comment (kbhawkey) [inline content/en/docs/setup/production-environment/dockershim.md:28]: Can you reduce the text in this heading? I have not viewed the rendered page, but this heading will display in the right hand TOC and the text will wrap over multiple lines?

File: content/en/docs/setup/production-environment/dockershim.md [hunk 26]
```diff
@@ -0,0 +1,170 @@
...
+- Other support scripts or daemons that expect docker to be available and are run
+  outside of Kubernetes (e.g. monitoring or security agents)
+- GPUs or special hardware and how they integrate with your runtime and Kubernetes
+
+If you use Kubernetes resource requests/limits or file-based log collection
+DaemonSets then they will continue to work the same, but if you’ve customized
+your dockerd configuration, you’ll need to adapt that for your new container
+runtime where possible.
+
+Another thing to look out for is anything expecting to run [ed: how to say
```
Comment (kbhawkey) [inline content/en/docs/setup/production-environment/dockershim.md:113]: This is hard to parse:
Another thing to look out for is anything expecting to run [ed: how to say

File: content/en/docs/setup/production-environment/dockershim.md [hunk 27]
```diff
@@ -0,0 +1,170 @@
...
+docker CLI commands in a way that isn’t repetitive] for system maintenance or
+nested inside a container when building images will no longer work. For the
+former, you can use the [`crictl`][cr] tool as a mostly drop-in replacement
```
Comment (kbhawkey) [inline content/en/docs/setup/production-environment/dockershim.md:116]: nit: remove or reword "as a mostly drop-in"

Comment (kbhawkey) [reply content/en/docs/setup/production-environment/dockershim.md:116]: nit: The links.

File: content/en/docs/setup/production-environment/dockershim.md [hunk 28]
```diff
@@ -0,0 +1,170 @@
...
+and for the latter you can use newer container build options like [img], [buildah],
+or [kaniko] that don’t require Docker.
+
+[cr]: https://github.com/kubernetes-sigs/cri-tools
+[img]: https://github.com/genuinetools/img
+[buildah]: https://github.com/containers/buildah
+[kaniko]: https://github.com/GoogleContainerTools/kaniko
+
+For containerd in particular, you can start with their [documentation] to see what
+configuration options are available as you migrate things over.
```
Comment (kbhawkey) [inline content/en/docs/setup/production-environment/dockershim.md:126]: nit: could remove "in particular" and make the link text more descriptive

**Comment by kbhawkey** [comment]
Page preview:
https://deploy-preview-25360--kubernetes-io-master-staging.netlify.app/docs/setup/production-environment/dockershim/

Not clear on this location, but I would place this page right after the Container runtimes page. Check the weight of the Container runtimes page.

**Comment by celestehorgan** [comment]
@kbhawkey – I asked @mrbobbytables to move it to a blog post, because I'm not a fan of adding one-off/time sensitive FAqs.. There's quite a bit of discussion on that in #sig-contribex in slack, which he's done.

**Comment by mrbobbytables** [comment]
I think I addressed everything but the the nit on the links - I'm used the ref link format and didn't want to change all of them lol The ref link style is what we encourage in the contributor docs.

**Comment by celestehorgan** [comment]
/approve

I leave the lgtm to some other kind soul :)

**Comment by jeremyrickard** [comment]
/lgtm

Thank you @mrbobbytables

**Comment by BenTheElder** [comment]
/lgtm

thank you @mrbobbytables ! 😄
